src/shim.c

```diff
@@ -703,6 +704,9 @@ handle_proxy_message(struct cc_shim *shim)
 
 	fr = read_frame(shim);
 	if ( !fr) {
+		if (! reconnect_to_proxy(shim)) {
```

Comment:
You're making lots of assumption by calling this after you got a `fr == NULL`. Indeed, looking at `read_frame()` function, there are several reasons why we could get an error, and they are not all related to the proxy disconnection.